crypto/src/main/scala/org/bitcoins/crypto/Sign.scala

```diff
@@ -32,88 +29,164 @@ trait Sign {
+  private def asyncSignLowR(bytes: ByteVector, startAt: Long)(implicit
       ec: ExecutionContext): Future[ECDigitalSignature] = {
     val sigF: Future[ECDigitalSignature] = if (startAt == 0) {
-      signFunction(bytes)
+      asyncSign(bytes)
     } else {
       val startBytes = ByteVector.fromLong(startAt).padLeft(32).reverse
-      signWithEntropyFunction(bytes, startBytes)
+      asyncSignWithEntropy(bytes, startBytes)
     }
 
     sigF.flatMap { sig =>
       if (sig.bytes.length <= 70) {
         Future.successful(sig)
       } else {
-        signLowRFuture(bytes, startAt + 1)
+        asyncSignLowR(bytes, startAt + 1)
       }
     }
   }
 
-  def signLowRFuture(bytes: ByteVector)(implicit
+  def asyncSignLowR(bytes: ByteVector)(implicit
       ec: ExecutionContext): Future[ECDigitalSignature] = {
-    signLowRFuture(bytes, startAt = 0)
+    asyncSignLowR(bytes, startAt = 0)
+  }
+
+  def publicKey: ECPublicKey
+}
+
+object AsyncSign {
+
+  private case class AsyncSignImpl(
+      asyncSignFunction: ByteVector => Future[ECDigitalSignature],
+      asyncSignWithEntropyFunction: (
+          ByteVector,
+          ByteVector) => Future[ECDigitalSignature],
+      override val publicKey: ECPublicKey)
+      extends AsyncSign {
+
+    override def asyncSign(bytes: ByteVector): Future[ECDigitalSignature] = {
+      asyncSignFunction(bytes)
+    }
+
+    override def asyncSignWithEntropy(
+        bytes: ByteVector,
+        entropy: ByteVector): Future[ECDigitalSignature] = {
+      asyncSignWithEntropyFunction(bytes, entropy)
+    }
   }
 
-  def sign(bytes: ByteVector): ECDigitalSignature = {
-    Await.result(signFuture(bytes), 30.seconds)
+  def apply(
+      asyncSign: ByteVector => Future[ECDigitalSignature],
+      asyncSignWithEntropy: (
+          ByteVector,
+          ByteVector) => Future[ECDigitalSignature],
+      pubKey: ECPublicKey): AsyncSign = {
+    AsyncSignImpl(asyncSign, asyncSignWithEntropy, pubKey)
   }
 
-  def signLowR(bytes: ByteVector)(implicit
-      ec: ExecutionContext): ECDigitalSignature = {
-    Await.result(signLowRFuture(bytes), 30.seconds)
+  def constant(sig: ECDigitalSignature, pubKey: ECPublicKey): AsyncSign = {
+    AsyncSignImpl(_ => Future.successful(sig),
+                  (_, _) => Future.successful(sig),
+                  pubKey)
   }
 
+  /** This dummySign function is useful for the case where we do not have the
+    * signFunction available on the same jvm as the place where we are creating the
+    * sign. I can't think of a good way to serialize the signFunction, so it needs to be
+    * optional for now. Maybe we rethink the idea of the signFunction in the future.
+    * the public key is still useful here though because it can be used to match against
+    * a specific private key on another server
+    */
+  def dummySign(publicKey: ECPublicKey): AsyncSign = {
+    constant(EmptyDigitalSignature, publicKey)
+  }
+}
+
+trait Sign extends AsyncSign {
```

Comment:
Is this meant to extend `AsyncSign`? I thought the purpose is to be synchronous

Reply:
If some other function requires an `AsyncSign`, it is always possible to give it a `Sign` which as you'll see below simply wraps results in `Future.successful`.